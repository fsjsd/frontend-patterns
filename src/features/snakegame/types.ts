
/**
 * Represents coordinate on the board
 */
export interface Coord {
  x: number,
  y: number
}

/**
 * State shape for the game
 */
export interface SnakeGameShape {
  /**
   * 'Speed' of the snake, in milliseconds. null infers that the game is paused.
   */
  speedMs: number | null,
  /**
   * Current direction of the snake
   */
  direction: string,
  /**
   * Position of the Apple
   */
  applePos: Coord,
  /**
   * Snake's coordinates on the board
   */
  snakeCoords: Coord[];
  /**
   * Score for current round
   */
  score: number;
  /**
   * Score for current round
   */
  highScore: number;
} 

export enum SnakeDirection {
  ArrowUp = 'ArrowUp',
  ArrowDown = 'ArrowDown',
  ArrowLeft = 'ArrowLeft',
  ArrowRight = 'ArrowRight'
}

export enum SnakeGameOperation {
  Reset = 'Reset',
  Advance = 'Advance',
  Start = 'Start'
}

export type SnakeGameActionType = SnakeGameOperation | SnakeDirection;

export interface SnakeGameAction {
  type: SnakeGameActionType,
  payload?: unknown
}

export interface BoardViewProps {
  rows: number,
  cols: number,
  cells: number[][],
  state: SnakeGameShape
}


export enum ViewMode {
  Html = 'html',
  Ascii = 'ascii',
  Canvas = 'canvas'
}