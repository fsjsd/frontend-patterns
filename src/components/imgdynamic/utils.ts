// Putting this in it's own file so it can be mocked
export const NATIVE_LAZYLOAD_SUPPORTED = 'loading' in HTMLImageElement.prototype;