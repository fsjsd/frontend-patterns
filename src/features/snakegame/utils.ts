import { Coord, SnakeGameShape } from "./types";

export const BOARD_COLS = 21;
export const BOARD_ROWS = 21;

export const INITIAL_SNAKE_SPEED = 200;

export const INITIAL_SNAKE_COORDS: Coord[] = [
  { x: 10, y: 10 },
  { x: 10, y: 11 },
  { x: 10, y: 12 }
]


export const randomApplePos = (cols:number = BOARD_COLS, rows:number = BOARD_ROWS): Coord => {
  const x = Math.floor(Math.random() * cols);
  const y = Math.floor(Math.random() * rows);
  // TODO: Don't collide with existing snake coords
  // - Simple implementation would be load all snake coords
  //   into two Sets of x and y then while() on the random
  //   until a non-colliding coord is found.
  return { x, y };
}

export const isCoordEqual = (coord1: Coord, coord2: Coord) => {
  return coord1.x === coord2.x && coord1.y === coord2.y;
}

export const isSnakeCollision = (coord: Coord, snakeCoords: Coord[]) => {
  return snakeCoords.some(snakeCoord => isCoordEqual(snakeCoord, coord))
}


// implemented as function otherwise randomApplePos() invokes on declaration
// before unit tests can mock it out
export const initialState = (): SnakeGameShape => ({
  applePos: randomApplePos(),
  snakeCoords: INITIAL_SNAKE_COORDS,
  direction: 'ArrowUp',
  speedMs: null,
  score: 0,
  highScore: 0
})