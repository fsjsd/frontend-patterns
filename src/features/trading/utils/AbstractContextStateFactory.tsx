import React, { ReactNode } from 'react'

/*

const PasswordContext = React.createContext<string | null>(null)
const SetPasswordContext = React.createContext<React.Dispatch<
  React.SetStateAction<string>
> | null>(null)

function PasswordProvider({children}: {children: React.ReactNode}) {
  const [password, setPassword] = React.useState('')

  return (
    <PasswordContext.Provider value={password}>
      <SetPasswordContext.Provider value={setPassword}>
        {children}
      </SetPasswordContext.Provider>
    </PasswordContext.Provider>
  )
}

function usePassword() {
  const password = React.useContext(PasswordContext)
  if (password === null) {
    throw new Error(
      'usePassword must be used in a decendent of PasswordProvider',
    )
  }
  return password
}

function useSetPassword() {
  const setPassword = React.useContext(SetPasswordContext)
  if (setPassword === null) {
    throw new Error(
      'usePassword must be used in a decendent of PasswordProvider',
    )
  }
  return setPassword
}*/

// function useState<S>(initialState: S | (() => S)): [S, Dispatch<SetStateAction<S>>];

type AbstractContextStateFactoryResult<T> = [
  ({ value, children }: { value: T, children: React.ReactNode }) => JSX.Element,
  () => T,
  () => React.Dispatch<React.SetStateAction<T>>
];

function AbstractContextStateFactory<T>(): AbstractContextStateFactoryResult<T> {
  const ValueContext = React.createContext<T | null>(null)
  const SetValueContext = React.createContext<React.Dispatch<
    React.SetStateAction<T>
  > | null>(null)

  function AbstractProvider({ value, children }: { value: T, children: React.ReactNode }) {
    const [val, setVal] = React.useState<T>(value)

    return (
      <ValueContext.Provider value={val}>
        <SetValueContext.Provider value={setVal}>
          {children}
        </SetValueContext.Provider>
      </ValueContext.Provider>
    )
  }

  function useValue() {
    const password = React.useContext(ValueContext)
    if (password === null) {
      throw new Error(
        'use{Value} must be used in a decendent of Provider',
      )
    }
    return password
  }

  function useSetValue() {
    const setValue = React.useContext(SetValueContext)
    if (setValue === null) {
      throw new Error(
        'useSet{Value} must be used in a decendent of Provider',
      )
    }
    return setValue
  }

  return [AbstractProvider, useValue, useSetValue];
}


export default AbstractContextStateFactory;