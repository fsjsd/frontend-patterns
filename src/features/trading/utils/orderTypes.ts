

type KeyOf<T extends object> = Extract<keyof T, string>;

type ValueOf<T> = T[keyof T];

export enum OrderSide {
  buy = 'buy',
  sell = 'sell'
}
export interface Order {
  security: string;
  side: OrderSide;
  limitprice: number;
  quantity: number;
}

export type OrderKey = KeyOf<Order>;
export type OrderValue = ValueOf<Order>;

export interface ValidationResult {
  valid: boolean;
  message?: string
}