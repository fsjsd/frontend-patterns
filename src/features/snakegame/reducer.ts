import { Coord, SnakeDirection, SnakeGameAction, SnakeGameOperation, SnakeGameShape } from "./types";
import { BOARD_COLS, BOARD_ROWS, initialState, INITIAL_SNAKE_COORDS, INITIAL_SNAKE_SPEED, isCoordEqual, isSnakeCollision, randomApplePos } from "./utils";

type ReducerMap = { 
  [key in SnakeDirection | SnakeGameOperation]: (state: SnakeGameShape) => SnakeGameShape
}

export const snakeGameReducer = (state: SnakeGameShape, action: SnakeGameAction) => {
  const reducerMap: ReducerMap = {
    [SnakeDirection.ArrowUp]: (state: SnakeGameShape) => {
      if(state.direction === SnakeDirection.ArrowDown) {
        return state;
      }
      return {
        ...state,
        direction: SnakeDirection.ArrowUp
      }
    },
    [SnakeDirection.ArrowDown]: (state: SnakeGameShape) => {
      if(state.direction === SnakeDirection.ArrowUp) {
        return state;
      }
      return {
        ...state,
        direction: SnakeDirection.ArrowDown
      }
    },
    [SnakeDirection.ArrowLeft]: (state: SnakeGameShape) => {
      if(state.direction === SnakeDirection.ArrowRight) {
        return state;
      }
      return {
        ...state,
        direction: SnakeDirection.ArrowLeft
      }
    },
    [SnakeDirection.ArrowRight]: (state: SnakeGameShape) => {
      if(state.direction === SnakeDirection.ArrowLeft) {
        return state;
      }
      return {
        ...state,
        direction: SnakeDirection.ArrowRight
      }
    },
    [SnakeGameOperation.Reset]: () => {
      return {
        ...initialState(),
        highScore: state.highScore
      }
    },
    [SnakeGameOperation.Start]: (state: SnakeGameShape) => {
      return {
        ...state,
        speedMs: INITIAL_SNAKE_SPEED
      }
    },
    [SnakeGameOperation.Advance]: (state: SnakeGameShape) => {
      let newSnakeCoords = state.snakeCoords.slice();
      const prevHead = state.snakeCoords[0];
      let newHead: Coord | null = null;
      let hitSomething = false;
    
      switch (state.direction) {
        case 'ArrowUp':
          newHead = { ...prevHead, y: prevHead.y - 1 };
          if (prevHead.y === 0) { //at border
            hitSomething = true; 
          } else {
            newSnakeCoords = [
              newHead,
              ...state.snakeCoords.slice(0, state.snakeCoords.length - 1)
            ]
          }
          break;
        case 'ArrowDown':
          newHead = { ...prevHead, y: prevHead.y + 1, };
          if (prevHead.y === BOARD_ROWS - 1 ) { //at border
            hitSomething = true;
          } else {
            newSnakeCoords = [
              newHead,
              ...state.snakeCoords.slice(0, state.snakeCoords.length - 1)
            ]
          }
          break;
        case 'ArrowLeft':
          newHead = { ...prevHead, x: prevHead.x - 1, };
          if (prevHead.x === 0) { //at border
            hitSomething = true;
          } else {
            newSnakeCoords = [
              newHead,
              ...state.snakeCoords.slice(0, state.snakeCoords.length - 1)
            ]
          }
          break;
        case 'ArrowRight':
          newHead = { ...prevHead, x: prevHead.x + 1, };
          if (prevHead.x === BOARD_COLS - 1) {
            hitSomething = true;
          } else {
            newSnakeCoords = [
              newHead,
              ...state.snakeCoords.slice(0, state.snakeCoords.length - 1)
            ]
          }
          break;
      }
    
      let newApplePos = state.applePos;
      let newSpeed: number | null = state.speedMs ?? INITIAL_SNAKE_SPEED;
      let score = state.score;
      let newDirection = state.direction;
      let newHighScore = state.highScore;
    
      // BUG: if snake reverses, will die
      if(newHead !== null && !hitSomething) { // micro optimisation if we know there's already a collision
        // this looks at all previous snake coords to detect collision, since newHead
        // will be in new cell
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        hitSomething = isSnakeCollision(newHead!, state.snakeCoords)
      }

      // Hits apple
      if (newHead && isCoordEqual(newHead, state.applePos)) {
        newSpeed = newSpeed * 0.9;
        // new applepos
        newApplePos = randomApplePos();
        // append old tail to extend tail
        newSnakeCoords = [
          ...newSnakeCoords,
          state.snakeCoords[state.snakeCoords.length - 1]
        ]
        // increase score
        score = score + 1;

        // increase high score
        if(score > state.highScore){
          newHighScore = score;
        }
      }

      if (hitSomething) {
        newSpeed = null;
        newApplePos = randomApplePos();
        newSnakeCoords = INITIAL_SNAKE_COORDS;
        score = 0;
        newDirection = SnakeDirection.ArrowUp;
      }
    
      return {
        speedMs: newSpeed,
        direction: newDirection,
        applePos: newApplePos,
        snakeCoords: newSnakeCoords,
        score,
        highScore: newHighScore
      };
    },
  }

  if(reducerMap[action.type]) {
    return reducerMap[action.type](state);
  }
  
  return state;
};