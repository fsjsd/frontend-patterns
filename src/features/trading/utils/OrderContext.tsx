import AbstractContextStateFactory from "./AbstractContextStateFactory";
import { Order } from "./orderTypes";


const [OrderProvider, useOrder, useSetOrder] = AbstractContextStateFactory<Order>();

export {
  OrderProvider,
  useOrder,
  useSetOrder
}