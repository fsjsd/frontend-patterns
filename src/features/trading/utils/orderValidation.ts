import { OrderSide, ValidationResult } from "./orderTypes";

const validations = {
  security: (value: string): ValidationResult => {
    if(value === undefined || value === '') {
      return {
        valid: false,
        message: 'Security required'
      };
    }
    return { valid: true };
  },
  limitprice: (value: string): ValidationResult => {
    if(value === undefined || value === '') {
      return {
        valid: false,
        message: 'Limit price required'
      };
    }
    const limitprice = parseFloat(value);
    if(limitprice < 0) {
      return {
        valid: false,
        message: 'Limit price must be greater than 0'
      };
    }
    return { valid: true };
  },
  quantity: (value: string, side: OrderSide, availableFunds: number, securityPrice: number, holdings?: number): ValidationResult => {
    if(value === undefined || value === '') {
      return {
        valid: false,
        message: 'Quantity required'
      };
    }
    const quantity = parseFloat(value);
    if(quantity < 0) {
      return {
        valid: false,
        message: 'Quantity must be greater than 0'
      };
    }
    if(holdings !== undefined && side === OrderSide.sell && quantity > holdings) {
      return {
        valid: false,
        message: 'Quantity exceeds holdings'
      };
    }
    if(holdings !== undefined && side === OrderSide.buy && (quantity * securityPrice) > availableFunds) {
      return {
        valid: false,
        message: 'Quantity exceeds available funds'
      };
    }
    return { valid: true };
  }
}

export default validations;

/*

function validateOrderField (key: OrderKey, value: string | undefined) : ValidationResult {
  if(key === 'limitprice') {
    if(value === undefined || value === '') {
      return {
        valid: false,
        message: 'Limit price required'
      };
    }
    const limitprice = parseFloat(value);
    if(limitprice < 0) {
      return {
        valid: false,
        message: 'Limit price must be greater than 0'
      };
    }
  }
  return { valid: true };
}


export default validateOrderField;
*/