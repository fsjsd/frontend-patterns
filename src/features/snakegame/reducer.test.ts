import { SnakeDirection, SnakeGameOperation, SnakeGameShape } from "./types";
import { snakeGameReducer } from "./reducer";
import { BOARD_COLS, BOARD_ROWS, initialState, INITIAL_SNAKE_SPEED } from "./utils";

describe("snakeGameReducer", () => {
  beforeEach(() => {
    jest.spyOn(global.Math, "random").mockReturnValue(0.123456789);
  });
  afterEach(() => {
    jest.spyOn(global.Math, "random").mockRestore();
  });

  test("start sets interval speed", () => {
    const state = initialState();
    expect(state.speedMs).toEqual(null);
    expect(
      snakeGameReducer(state, { type: SnakeGameOperation.Start }).speedMs
    ).toEqual(INITIAL_SNAKE_SPEED);
  });
  
  test("should return the initial state", () => {
    expect(
      snakeGameReducer(initialState(), { type: SnakeGameOperation.Reset })
    ).toEqual(initialState());
  });

  test("should preserve high score after game reset", () => {
    expect(
      snakeGameReducer(
        {
          ...initialState(),
          highScore: 3,
        },
        { type: SnakeGameOperation.Reset }
      )
    ).toEqual({
      ...initialState(),
      highScore: 3,
    });
  });

  test.each([
    [SnakeDirection.ArrowUp, SnakeDirection.ArrowDown, SnakeDirection.ArrowUp],
    [
      SnakeDirection.ArrowUp,
      SnakeDirection.ArrowLeft,
      SnakeDirection.ArrowLeft,
    ],
    [
      SnakeDirection.ArrowDown,
      SnakeDirection.ArrowUp,
      SnakeDirection.ArrowDown,
    ],
    [
      SnakeDirection.ArrowDown,
      SnakeDirection.ArrowLeft,
      SnakeDirection.ArrowLeft,
    ],
    [
      SnakeDirection.ArrowLeft,
      SnakeDirection.ArrowRight,
      SnakeDirection.ArrowLeft,
    ],
    [SnakeDirection.ArrowLeft, SnakeDirection.ArrowUp, SnakeDirection.ArrowUp],
    [
      SnakeDirection.ArrowRight,
      SnakeDirection.ArrowLeft,
      SnakeDirection.ArrowRight,
    ],
    [SnakeDirection.ArrowRight, SnakeDirection.ArrowUp, SnakeDirection.ArrowUp],
  ])(
    "reverse direction is not allowed %o -> %o",
    async (currentDirection, intendedDirection, result) => {
      expect(
        snakeGameReducer(
          {
            ...initialState(),
            direction: currentDirection,
          },
          { type: intendedDirection }
        ).direction
      ).toEqual(result);
    }
  );

  test.each([
    [SnakeDirection.ArrowLeft, SnakeDirection.ArrowUp],
    [SnakeDirection.ArrowLeft, SnakeDirection.ArrowDown],
    [SnakeDirection.ArrowUp, SnakeDirection.ArrowLeft],
    [SnakeDirection.ArrowUp, SnakeDirection.ArrowRight],
  ])(
    "direction change moves snake - %o -> %o",
    async (currentDirection, intendedDirection) => {
      const stateStart = {
        ...initialState(),
        direction: currentDirection,
      };
      const stateAfterDirectionChange = snakeGameReducer(stateStart, {
        type: intendedDirection,
      });
      const stateAfterAdvance = snakeGameReducer(stateStart, {
        type: SnakeGameOperation.Advance,
      });

      expect(stateAfterDirectionChange.direction).toEqual(intendedDirection);
      expect(stateAfterAdvance.snakeCoords).toMatchSnapshot();
    }
  );

  test("apple collision wins round", () => {
    const state: SnakeGameShape = {
      ...initialState(),
      snakeCoords: [
        { x: 2, y: 2 },
        { x: 2, y: 1 },
        { x: 2, y: 0 },
      ],
      applePos: { x: 2, y: 3 },
      direction: SnakeDirection.ArrowDown,
    };
    expect(state.score).toEqual(0);
    expect(state.highScore).toEqual(0);

    const stateAfterAdvance = snakeGameReducer(state, {
      type: SnakeGameOperation.Advance,
    });
    expect(stateAfterAdvance.snakeCoords).toEqual([
      { x: 2, y: 3 },
      { x: 2, y: 2 },
      { x: 2, y: 1 },
      { x: 2, y: 0 },
    ]);
    expect(stateAfterAdvance.score).toEqual(1);
    expect(stateAfterAdvance.highScore).toEqual(1);
  });

  test.each([
    [SnakeDirection.ArrowUp, { x: 2, y: 0 }, { x: 2, y: 1 }],
    [SnakeDirection.ArrowLeft, { x: 0, y: 1 }, { x: 1, y: 1 }],
    [SnakeDirection.ArrowDown, { x: 2, y: BOARD_ROWS-1 }, { x: 2, y: BOARD_ROWS }],
    [SnakeDirection.ArrowRight, { x: BOARD_COLS-1, y: 1 }, { x: BOARD_COLS, y: 1 }],
  ])("round ends if boundary hit - %o", (direction, head, tail) => {
    const state: SnakeGameShape = {
      ...initialState(),
      speedMs: 200,
      snakeCoords: [
        head,
        tail,
      ],
      direction: direction,
    };

    const stateAfterAdvance = snakeGameReducer(state, {
      type: SnakeGameOperation.Advance,
    });
    // speed reset after round end / collision
    expect(stateAfterAdvance.speedMs).toBeNull();
    // snake reset
    expect(stateAfterAdvance.snakeCoords).toEqual(initialState().snakeCoords);
  });
});
