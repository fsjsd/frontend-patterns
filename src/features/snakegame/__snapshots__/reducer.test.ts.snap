// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snakeGameReducer direction change moves snake - 'ArrowLeft' -> 'ArrowDown' 1`] = `
Array [
  Object {
    "x": 9,
    "y": 10,
  },
  Object {
    "x": 10,
    "y": 10,
  },
  Object {
    "x": 10,
    "y": 11,
  },
]
`;

exports[`snakeGameReducer direction change moves snake - 'ArrowLeft' -> 'ArrowUp' 1`] = `
Array [
  Object {
    "x": 9,
    "y": 10,
  },
  Object {
    "x": 10,
    "y": 10,
  },
  Object {
    "x": 10,
    "y": 11,
  },
]
`;

exports[`snakeGameReducer direction change moves snake - 'ArrowUp' -> 'ArrowLeft' 1`] = `
Array [
  Object {
    "x": 10,
    "y": 9,
  },
  Object {
    "x": 10,
    "y": 10,
  },
  Object {
    "x": 10,
    "y": 11,
  },
]
`;

exports[`snakeGameReducer direction change moves snake - 'ArrowUp' -> 'ArrowRight' 1`] = `
Array [
  Object {
    "x": 10,
    "y": 9,
  },
  Object {
    "x": 10,
    "y": 10,
  },
  Object {
    "x": 10,
    "y": 11,
  },
]
`;
